src/core/node_clone.cc

```diff
@@ -0,0 +1,112 @@
+/*
+ *  OpenSCAD (www.openscad.org)
+ *  Copyright (C) 2009-2011 Clifford Wolf <clifford@clifford.at> and
+ *                          Marius Kintel <marius@kintel.net>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  As a special exception, you have permission to link this program
+ *  with the CGAL library and distribute executables, as long as you
+ *  follow the requirements of the GNU GPL in regard to all of the
+ *  software in the executable aside from CGAL.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include "src/geometry/linalg.h"
+#include "src/geometry/GeometryUtils.h"
+#include "src/core/primitives.h"
+#include "src/core/TransformNode.h"
+#include "src/core/RotateExtrudeNode.h"
+#include "src/core/LinearExtrudeNode.h"
+#include "src/core/CgalAdvNode.h"
+#include "src/core/CsgOpNode.h"
+#include "src/core/ColorNode.h"
+#include "src/core/RoofNode.h"
+#include "src/core/RenderNode.h"
+#include "src/core/SurfaceNode.h"
+#include "src/core/TextNode.h"
+#include "src/core/OffsetNode.h"
+#include "src/core/ProjectionNode.h"
+#include "src/core/ImportNode.h"
+
+std::vector<ModuleInstantiation *> modinsts_list;
+
+#define NodeCloneFunc(T) std::shared_ptr<T> clone_what(const T *node) {\
+       	ModuleInstantiation *inst = new ModuleInstantiation(node->modinst->name() ,\
```

Comment:
@gsohler Reviewing some old code: This is a naked pointer which is (I believe) never deleted. Nodes generally keep a const* to the ModuleInstantiation around for read-only purposes. The ModuleInstantiation is owned by the instantiating scope and we promise that the scope objects outlive the node tree. Could we instead just copy the pointer here?